### ISSUE TITLE
Provide an IAmAMessageConsumerAsync interface

### ISSUE THREAD

**Issue by SeanFarrow** [BrighterCommand/Brighter#308]
It would be nice to have an IAmAMessageConsumerAsync interface as a companion to the IAmAMessageProducerAsync interface.
Given that a lot of messaging gateways are using the tpl, this makes perfect sense to me.
I'm happy to do the work if desired.

**Comment by iancooper** [comment]
@SeanFarrow We have thought about this a little bit, and its not quite as simple as it sounds. First our consumers use a single-threaded message pump via a supervisor worker pattern. These are long-running threads. If you need to scale out, you increase the number of performers i.e. threads. This is already parallel. Normally you don't instantiate a consumer directly, but use the dispatcher instead, which returns a Task anyway.

In principle we block on I/O to the message queue. We do want to look at that but it relates to another issue, which is what would it mean for that thread to be async whilst it waited on a new message. In essence you have blocked the pump anyway, because if there is no message to consume, there is theoretically no work for that thread to do. Now you could yield, but the O/S scheduler coupled with a timeout would do that anyway. We do have some work to do here to move back to a consumer model on RMQ to allow us to create a local buffer for multiple performers that may ease this.

We could create an async pipeline to support async handlers. but we have an issue with deadlocks if we don't callback on a pool thread, and that breaks anyone who needs the continuation to run in the same context. So we are looking for solutions to that, which possibly means that the pump runs on a new thread, but that would destroy ordering etc.

So, whilst we have looked at this, and still look at this, there are no obvious answers that actually help the client. We may revisit once I have done buffered queues.

**Comment by SeanFarrow** [comment]
@IanCooper,
That all makes ssense.
Is there any chance of an EventStore gateway? I’m not sure whether I raised a separate issue for this? What do you do in terms of checkpoints, so that messages are not processed more than once in gateways where the last message position isn’t stored?

**Comment by iancooper** [comment]
@SeanFarrow We had Event Store on our backlog for a long time, and someone keen to write it, but it never happened. I'm open to offers. I might get to it, if someone else does not first.

Currently we use the Command Sourcing attribute with OnceOnly to de-duplicate i.e. I have seen this before. We have some fixes at Huddle (on a branch of that attribute) that I am pushing to get the author to PR. Mainly it involves moving the write to the command store until after the target handler has been processed, so that you can replay failed messages.

**Comment by SeanFarrow** [comment]
@IanCooper,
That makes sense. What would be your time frame for this to be PR’d ideally?

It would be good to have a message gateway, the only question I would have is how would you want to acknowledge messages, as without using the persistent subscriptions feature of EventStore, there is no built-in way of doing this.
What is the minimum needed to write a message gateway in terms of classes/interface implementations?

**Comment by SeanFarrow** [comment]
@IanCooper,
I've just looked at doing this, based on the existing transports. It looks straightforward, but EventStore don't have timeouts on individual reads, they just have a timeout on the connection for all reads which is disposable, so I wouldn't want to create a connection per operation. I know the sql gateway does this, so am happy to if it's the best option.
How do you handle async calls in sync code? I'm assuming you are just calling wait on the returned task?

**Comment by iancooper** [comment]
I'm traveling, let me look tomorrow.

**Comment by iancooper** [comment]
Going to close this for now. We know what we need to do to get the STA pump to work with async, though it has side-affects (de-ordering messages)
